Deprecate requirements format "base=>1.0[extra]"

jku opened the pull request:
Note: I'm new to pip code and it was not obvious where this should be done: feel free to point me to another direction. I chose this spot in `parse_req_from_line()` because there's other validation right next to it. 

There's currently no "gone_in" defined for the deprecation: let me know if there's a policy I should follow here.

The bug also has my comments on some other weirdness that may or may not be related bugs.

Fixes #8288

--- 
This requirements format does not conform to PEP-508. Add deprecation
warning and fix the corresponding test.

Note that the actual added check ensures the specifier is compliant
with PEP-440 (the specifier version is invalid "1.0[extra]").

pradyunsg commented on src/pip/_internal/req/constructors.py:
```diff
@@ -353,6 +354,12 @@ def with_source(text):
     if req_as_string is not None:
         try:
             req = Requirement(req_as_string)
+            # Check that specifiers are PEP-440 compliant
+            for spec in req.specifier:
+                if not isinstance(spec, Specifier):
+                    msg = "Invalid specifier format '{}'".format(str(spec))
+                    deprecated(msg, replacement=None, gone_in=None)
```
We should probably move this entire loop into an else block associated with this `try-except`.

pradyunsg replied:
I think:

- `replacement` should be "use PEP 440-compatible specifiers. <pep 440 link>" (or similar).
- `gone_in` should be `"21.0"`.

jku commented:
thanks for review @pradyunsg, I've done the requested changes and a minor tweak to the actual message.

uranusjr commented on src/pip/_internal/req/constructors.py:
```diff
@@ -368,6 +369,14 @@ def with_source(text):
+        else:
+            # Check that specifiers are PEP-440 compliant
+            for spec in req.specifier:
+                if not isinstance(spec, Specifier):
```
What would the type of `spec` be, if not `Specifier`? Downstream distributors of pip perform “debundling” that changes the import path of `pip._vendor.packaging`, so we want to avoid `isinstance()` checks if possible, since they may return wrong results.

jku replied:
spec is either a Specifier (if it parses as PEP-440 specifier) or a LegacySpecifier (if it does not). Based on Changelogs this has been part of the packaging API since 2014 (packaging 14.3) but probably my use of SpecifierSet iteration only came available in 2015 (15.2).

I'm happy to use another method as well if you don't trust results from packaging.

pfmoore commented on tests/functional/test_new_resolver.py:
```diff
@@ -228,6 +226,32 @@ def test_new_resolver_installs_extras(tmpdir, script, root_dep):
     assert_installed(script, base="0.1.0", dep="0.1.0")
 
 
+def test_new_resolver_installs_extras_deprecated(tmpdir, script):
```
Why is this test in `test_new_resolver.py`? It's not explicitly resolver-related, is it? It should probably go in the main test suite.

pfmoore replied:
Also, do you not need to test use of the syntax on the command line, as well as in a requirements file?

jku replied:
command line test should probably go in test/functional/test_install_extras.py... I 'll do that

As to the resolver-related question: It looks like I added a test but really it's an existing test I moved from one function to a new one. Many of the resolver wheel-install tests do look like generic tests (at least to an outside observer) but maybe moving those is another PR?

pfmoore replied:
Ah, right. It's fine here in that case - yes, these are pretty generic tests, and at some point will need to be moved into the main test suite. That's why I was concerned about adding more here that we didn't need to.

But your logic's sound. Leave it here and we can move it when we tidy up the others.

pfmoore commented:
If I'm reading the PR correctly, we're actually deprecating *all* non-PEP 440 version numbers, not just the `1.0[extra]` style. But it's not the pip user that's in control of the version numbering used by the projects they are installing, so this has the potential to (in effect) make certain packages not installable with pip.

I'm not against this in principle, but have we checked how many packages this will apply to? Does PyPI enforce PEP 440 version format? I'd be fine with only considering PyPI here (private indexes can police themselves in this matter).

jku commented:
> If I'm reading the PR correctly, we're actually deprecating _all_ non-PEP 440 version numbers, not just the `1.0[extra]` style. But it's not the pip user that's in control of the version numbering used by the projects they are installing, so this has the potential to (in effect) make certain packages not installable with pip.

Fair point. I think having a look through pypi is prudent in any case... I'll think about this

uranusjr commented:
If I’m understanding the situation correctly, `base>=1.0[add]` is actually being parsed as:

* Name: `base`
* Version specifier: `>=1.0[add]` (which is a legacy specifier)

And somehow this is would later be used to select extra `add`?

If this is a case, the fix likely should not be to deprecate the legacy specifiers, but to figure out why it’s selecting the extra, and make it not do that.

jku commented:
> If I’m understanding the situation correctly, `base>=1.0[add]` is actually being parsed as:
> 
> * Name: `base`
> * Version specifier: `>=1.0[add]` (which is a legacy specifier)
> 
> And somehow this is would later be used to select extra `add`?

Correct. I tried to start a discussion on this in the bug, please see my last two comments in https://github.com/pypa/pip/issues/8288

> If this is a case, the fix likely should not be to deprecate the legacy specifiers, but to figure out why it’s selecting the extra, and make it not do that.

Well, it might still be prudent to warn or stop processing as early as possible -- but let's figure out the big picture first in any case: I think the issue discussion might be the right place for that

jku commented:
> If I'm reading the PR correctly, we're actually deprecating _all_ non-PEP 440 version numbers, not just the `1.0[extra]` style. But it's not the pip user that's in control of the version numbering used by the projects they are installing, so this has the potential to (in effect) make certain packages not installable with pip.
> 
> I'm not against this in principle, but have we checked how many packages this will apply to? Does PyPI enforce PEP 440 version format? I'd be fine with only considering PyPI here (private indexes can police themselves in this matter).

pypi does enforce pep-440: upload fails with non-conformant version.

I've had a look at wheel filenames in pypi repo: out of 1.1 million wheel files 84 have non-conforming versions (and are not somehow otherwise broken). 9 of them are current versions.  Something prevents installing them without the version: e.g. `pip install HolyGrail` fails with _No matching distribution found_ but `pip install HolyGrail==0.2.1.Perceval` succeeds.

So there is a small number of packages with non-conforming version numbers that currently are somewhat installable from pypi -- they are not likely to be very popular as the "non-versioned" installation method already fails.

pfmoore commented:
Thanks for doing that analysis. It sounds like *if* we decide to deprecate, the impact will be sufficiently low to make it a reasonable option (we'd still need the deprecation period, to cover non-PyPI projects, of course).

jku commented:
Changes since the previous review:
 * added test_install_deprecated_extra() to test command line
 * changed what is actually deprecated based on feedback: now it's only warning on any version string that ends with ']' (as a proxy to finding misparsed version strings that contains [extras]) instead of warning on every legacy version specifier. No deprecating LegacySpecifiers means there may be other cases like this ... but there may also be side-effects to not allowing them. _Feel free to tell me to switch this back if you think LegacySpecifiers should be deprecated instead -- I don't quite feel qualified to decide this_

> If this is a case, the fix likely should not be to deprecate the legacy specifiers, but to figure out why it’s selecting the extra, and make it not do that.

I don't think this is an option (as long as there is no refactoring of the whole function). We know why this happens: the version string is wrong so `Requirement()` fails to parse extras or the version properly but `_strip_extras()` happens store the extras by accident and the buggy version specifier is close enough to often work. The parsing in `Requirement()` is correct -- it just does not complain about using LegacySpecifiers. Before calling Requirement() we cannot reliably know that the version is fishy.

I do not think we should add _more_ parsing code to figure it out earlier  in `parse_req_from_line()` either: the correct path would probably be to try to handle the PEP-440 compatible requirement strings first and then add cases for the other requirements string formats (directory, file, url) -- possibly forming a PEP-440 compliant requirement string as result -- as well as any workarounds for deprecations and such. But as discussed in the bug that's likely to be a much bigger job.

jku commented:
Interesting, it looks like the command line test I added fails on windows: the package does not get created (alternatively, I do something wrong in the test). 

There's no  stdout/stderr listed for the AssertionError (not sure why, it does happen locally) but it's clear from the "created" list that the `simple` package does not get installed on windows. I think this is not related to my changes but can't be sure. I guess I'll have to open another PR to add just the test to confirm that (I don't have easy ways to test on windows)  

https://dev.azure.com/pypa/pip/_build/results?buildId=25874&view=logs&j=d5516b16-d635-5f23-ed1f-618b454a3b40&t=4b6551ce-09cf-548f-8bd5-83779dcb7360&l=90

uranusjr commented:
You’re hit by a quirk in pip’s test utilities. `pip install base>=1.0[extra]` is being interpreted as `pip install base` with the result being piped into a file named `=1.0[extra]`. The only good solution (aside from actually fixing the utility) is to use another operator…

uranusjr commented:
Or alterntively you can write the requirement line into a file and use `-r`, although I’m not sure if that would hit the code path you want to test (pip’s requirement parsing logic is complicated).

jku commented:
Thank you, that seems obvious now that you pointed it out.

This was a good find though: I had not extensively tested what actually passes the version check and I really should have... As I mentioned the specifier version becomes garbled: The tests I had picked all worked but it turns out quite a few of the comparisons do not, e.g. these do not work at all currently (examples from the test in question, requires_simple_extra == 0.1):

* `requires_simple_extra==0.1[extra]` fails to find a version because
* `requires_simple_extra<=0.1[extra]` fails to find a version because

I wonder if the right choice is actually to error out without a deprecation -- this seems to be quite thoroughly broken right now.

jku commented:
To summarize the slightly meandering thread, current state of the PR is:
* the last added test fails on windows so this should **not** be merged. I could workaround this but then...
* I realized this requirements format already fails for common version specifiers like  "base==0.1[extra]"

I think the right thing to do might be to just error out with the current check (does version end in "]"?) instead of deprecating. I can do one more pass of this if that sounds good (although how useful this is without a refactoring of `parse_req_from_line()` is debatable).

pradyunsg commented on tests/functional/test_install_extras.py:
```diff
@@ -104,6 +104,21 @@ def test_nonexistent_options_listed_in_order(script, data):
+
+    result = script.pip(
+        'install', '--no-index', '--find-links=' + data.find_links,
+        'requires_simple_extra>=0.1[extra]', expect_stderr=True,
```
Put this in a requirements.txt file and install from it.

We're hitting a bug in the test pipeline in Windows.

jku commented:
latest push changes `test_install_deprecated_extra()` only: it now uses a requirements file to avoid the shell redirect testing issue

jku commented:
latest push changes 
* test_install_deprecated_extra(): it now uses a requirements file to avoid the shell redirect testing issuelatest
* test_new_resolver_installs_extras_deprecated(): use --use-feature=2020-resolver

pradyunsg commented:
Thanks @jku! ^>^
